parsers/GT.py

```diff
@@ -1,163 +1,206 @@
-        else target_datetime.astimezone(TIMEZONE)
+        else target_datetime.astimezone(timezone.utc)
     )
-
-    api_data = _get_api_data(session, URL, target_datetime, target=ApiKind.PRODUCTION)
-
-    day_datetime = datetime.combine(  # truncate to day
-        target_datetime, time(), tzinfo=TIMEZONE
+    api_data = _get_api_data(
+        session,
+        target_datetime=target_datetime,
+        target=ApiKind.PRODUCTION,
+        num_backlog_days=0,
     )
-    results = [
-        {
-            "datetime": day_datetime + timedelta(hours=hour),
-            "production": {
-                "biomass": row["BIOGAS"] + row["BIOMASA"],
-                "coal": row["CARBÓN"],
-                "gas": row["GAS NATURAL"],
-                "geothermal": row["VAPOR"],
-                "hydro": row["AGUA"],
-                "oil": row["BUNKER"] + row["DIESEL"],
-                "solar": row["IRRADIACIÓN"],
-                "unknown": row["BIOMASA/CARBÓN"]
-                + row["CARBÓN/PETCOKE"]
-                + row["SYNGAN"],
-                "wind": row["VIENTO"],
-            },
-            "source": DOMAIN,
-            "zoneKey": zone_key,
-        }
-        for hour, row in enumerate(api_data)
-    ]
-    return [
-        validation.validate(
-            result, logger, floor=PRODUCTION_THRESHOLD, remove_negative=True
-        )
-        for result in results
-    ]
-
 
-def _get_api_data(
-    session: Session | None,
-    url: str,
-    date_time: datetime,
-    target: ApiKind,
-) -> list[dict]:
-    """Get the JSON-formatted response from the AMM API for the desired date-time."""
-    session = session or Session()
-    response_payload = session.get(
-        url, params={"dt": date_time.strftime("%d/%m/%Y")}
-    ).json()
-
-    # The JSON data returned by the API is a list of objects, each
-    # representing one technology type. Collect this information into a list,
-    # with the list index representing the hour of day.
-    results = [collections.defaultdict(float) for _ in range(24)]
-    for row in response_payload:
-        # The API returns hours in the range [1, 24], so each one refers to the
-        # past hour (e.g., 1 is the time period [00:00, 01:00)). Shift the hour
-        # so each index represents the hour ahead and is in the range [0, 24),
-        # e.g., hour 0 represents the period [00:00, 01:00).
-        results[int(row["hora"]) - 1][row["tipo"]] = row["potencia"]
-
-    is_historical_data = date_time.date() < datetime.now(TIMEZONE).date()
-
-    # For live consumption data, an hour's data isn't updated until the hour has passed,
-    # so the current (and future) hour(s) should not be included in the results.
-    # For live production data, the API will return zero-filled future data until the end of the day,
-    # so future hours should not be included in the results.
-    cutoff_index = (
-        date_time.hour if target == ApiKind.CONSUMPTION else date_time.hour + 1
-    )
```

Comment:
fix: I found out that this wasn't quite true and that an hour's data not being updated until the hour has passed applied to any kind of data - not only to consumption data. This resulted in getting production data whose last datapoint was still temporary data

For example, a production request made at 9:30 UTC had this (in-progress) datapoint for the 9->10 block: 
```
{'datetime': datetime.datetime(2024, 4, 10, 9, 0, tzinfo=datetime.timezone.utc), 'zoneKey': 'GT', 'production': {'biomass': 304.24, 'coal': 202.9, 'gas': 15.386, 'geothermal': 35.449, 'hydro': 95.92, 'oil': 173.165, 'solar': 0.0, 'unknown': 440.2, 'wind': 1.69}, 'storage': {}, 'source': 'wl12.amm.org.gt', 'sourceType': <EventSourceType.measured: 'measured'>, 'correctedModes': []}
```

while the same request made at 10:30 UTC had (completed) updated values:
```
{'datetime': datetime.datetime(2024, 4, 10, 9, 0, tzinfo=datetime.timezone.utc), 'zoneKey': 'GT', 'production': {'biomass': 305.42683, 'coal': 203.60906, 'gas': 11.92, 'geothermal': 33.34, 'hydro': 92.37983, 'oil': 188.9405, 'solar': 0.0, 'unknown': 440.72003, 'wind': 0.06}, 'storage': {}, 'source': 'wl12.amm.org.gt', 'sourceType': <EventSourceType.measured: 'measured'>, 'correctedModes': []}, 
```

The new code should cut things off more correctly (i.e. always cutoff the in-progress hour of live data)

Reply:
Nice!
As long as the datetimes are the same it causes no long term issues but very nice to have it fixed regardless!